parquet/file: NewParquetWriter panics on transient sink.Write errors during file initialization

### Describe the bug, including details regarding any error messages, version, and platform.

`(*file.Writer).startFile()` panics with `"failed to write magic number"` when the underlying sink's first `Write` returns an error or short-writes. Because `NewParquetWriter` calls `startFile` synchronously inside the constructor — and `NewParquetWriter` has no error return — there is no way for a consumer to recover via the standard Go error handling pattern.

**Code** ([`parquet/file/file_writer.go:201-204`](https://github.com/apache/arrow-go/blob/main/parquet/file/file_writer.go#L201-L204)):

```go
n, err := fw.sink.Write(magic)
if n != 4 || err != nil {
    panic("failed to write magic number")
}
```

**Constructor signature** ([`parquet/file/file_writer.go:72`](https://github.com/apache/arrow-go/blob/main/parquet/file/file_writer.go#L72)):

```go
func NewParquetWriter(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) *Writer
```

The higher-level [`pqarrow.NewFileWriter`](https://github.com/apache/arrow-go/blob/main/parquet/pqarrow/file_writer.go#L135) does have an `error` return, but it propagates the panic from `file.NewParquetWriter` through unconverted — so callers of either constructor must wrap the call in `defer recover()` to handle a flaky sink.

### Why this matters in practice

When the `sink` is backed by a network-attached writer — e.g., a `*storage.Writer` from `cloud.google.com/go/storage`, a Cloud Storage resumable upload, an HDFS client, an S3 multipart upload — a transient network blip on the first 4-byte write reliably crashes the consumer process.

Concrete example: a Kubernetes-deployed Go service that writes ~thousands of Parquet files per day to GCS sees ~1-2 pod crashes per 5 pod-hours under normal cluster conditions, all from this panic. Each crash:

1. Loses the entire worker process (not just the failing call)
2. Forces a container restart
3. Leaves any in-flight resumable upload session orphaned in GCS until session timeout
4. Requires application-level orphan-recovery machinery to reclaim work

The panic stack consistently looks like:

```
panic: failed to write magic number

goroutine N [running]:
github.com/apache/arrow-go/v18/parquet/file.(*Writer).startFile(...)
	github.com/apache/arrow-go/v18@v18.6.0/parquet/file/file_writer.go:203 +0x257
github.com/apache/arrow-go/v18/parquet/file.NewParquetWriter(...)
	github.com/apache/arrow-go/v18@v18.6.0/parquet/file/file_writer.go:90 +0x2d4
github.com/apache/arrow-go/v18/parquet/pqarrow.NewFileWriter(...)
	github.com/apache/arrow-go/v18@v18.6.0/parquet/pqarrow/file_writer.go:166 +0x34d
<consumer code calling pqarrow.NewFileWriter>
```

Workarounds exist (`defer recover()` at the call site) but they're brittle:
- Tied to an undocumented panic-string format
- Forced on every consumer of the public API
- Leave any partial sink state (open upload session, buffered bytes) for the caller to clean up

### Expected behavior

Per Go's errors-as-values convention and the contract suggested by `pqarrow.NewFileWriter`'s `(*FileWriter, error)` signature, a failure to write the magic header should surface as an error to the caller, not a panic.

### Suggested fixes (smallest change first)

1. **Add a sibling constructor** that returns `(*Writer, error)`, e.g.:

   ```go
   func NewParquetWriterWithError(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) (*Writer, error)
   ```

   Non-breaking. Existing callers keep current panic semantics; new callers opt into clean error handling. `pqarrow.NewFileWriter` switches to the new constructor internally so its `error` return becomes meaningful for this failure mode.

2. **Defer `startFile()` until first append.** Move the magic-header write to lazy execution on the first `AppendRowGroup` / `AppendBufferedRowGroup` call — both of which already have error paths. Slightly larger change but doesn't add a new public symbol.

3. **Lazy error pattern.** Have `startFile` store its error in an unexported `Writer.initErr` field, and have every other public method (`AppendRowGroup`, `Close`, `WritePageIndex`, …) short-circuit on it. Non-breaking but touches every public method.

### Reproduction sketch

```go
// Synthetic flaky sink that fails the first Write.
type flakySink struct{ writes int }

func (f *flakySink) Write(p []byte) (int, error) {
    f.writes++
    if f.writes == 1 {
        return 0, errors.New("synthesized network blip")
    }
    return len(p), nil
}

func main() {
    // ... build a trivial schema.GroupNode ...
    defer func() {
        if r := recover(); r != nil {
            log.Printf("recovered: %v", r) // prints "failed to write magic number"
        }
    }()
    _ = file.NewParquetWriter(&flakySink{}, schemaNode) // panics here
}
```

### Component(s)

Parquet

### Version

`github.com/apache/arrow-go/v18@v18.6.0` (also reproducible on current `main`; the [code at line 201-203](https://github.com/apache/arrow-go/blob/main/parquet/file/file_writer.go#L201-L204) is unchanged).

--- a/parquet/file/file_writer.go
+++ b/parquet/file/file_writer.go
@@ -70,12 +70,12 @@ func WithWriteMetadata(meta metadata.KeyValueMetadata) WriteOption {
 // If props is nil, then the default Writer Properties will be used. If the key value metadata is not nil,
 // it will be added to the file.
 //
-// This constructor panics with "failed to write magic number" if the
-// initial 4-byte write of the parquet magic header to the underlying sink
-// fails. Callers using a sink that may transiently fail (for example, a
-// cloud-storage upload writer or other network-attached writer) should
-// prefer [NewParquetWriterWithError], which surfaces the failure as a
-// returned error instead.
+// This constructor panics with the literal string "failed to write magic
+// number" if the initial write of the parquet magic header to the
+// underlying sink fails. The behavior is preserved for backward
+// compatibility with callers that string-match the panic value in a
+// recover() block; new code should prefer [NewParquetWriterWithError],
+// which returns the failure as an error instead.
 func NewParquetWriter(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) *Writer {
 	fw, err := NewParquetWriterWithError(w, sc, opts...)
 	if err != nil {
@@ -86,17 +86,16 @@ func NewParquetWriter(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) *W
 	return fw
 }
 
-// NewParquetWriterWithError is identical to [NewParquetWriter] except that
-// it returns an error to the caller if the initial write of the parquet
-// magic header to the underlying sink fails or short-writes. This is the
-// preferred constructor when the sink is backed by a writer that may
-// transiently fail (for example, a Google Cloud Storage resumable upload,
-// an S3 multipart upload, an HDFS client, or any other network-attached
-// writer) where a transient first-write failure should not panic the
-// calling process.
+// NewParquetWriterWithError returns a Writer that writes to the provided
+// io.Writer with the given schema.
 //
 // If props is nil, then the default Writer Properties will be used. If the
 // key value metadata is not nil, it will be added to the file.
+//
+// An error is returned if the initial write of the parquet magic header to
+// the underlying sink fails or short-writes, which can happen when the
+// sink is a flaky or network-attached writer (for example, a cloud-storage
+// upload writer).
 func NewParquetWriterWithError(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) (*Writer, error) {
 	config := &writerConfig{}
 	for _, o := range opts {

--- a/parquet/file/file_writer.go
+++ b/parquet/file/file_writer.go
@@ -65,11 +65,39 @@ func WithWriteMetadata(meta metadata.KeyValueMetadata) WriteOption {
 	}
 }
 
-// NewParquetWriter returns a Writer that writes to the provided WriteSeeker with the given schema.
+// NewParquetWriter returns a Writer that writes to the provided io.Writer with the given schema.
 //
 // If props is nil, then the default Writer Properties will be used. If the key value metadata is not nil,
 // it will be added to the file.
+//
+// This constructor panics with "failed to write magic number" if the
+// initial 4-byte write of the parquet magic header to the underlying sink
+// fails. Callers using a sink that may transiently fail (for example, a
+// cloud-storage upload writer or other network-attached writer) should
+// prefer [NewParquetWriterWithError], which surfaces the failure as a
+// returned error instead.
 func NewParquetWriter(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) *Writer {
+	fw, err := NewParquetWriterWithError(w, sc, opts...)
+	if err != nil {
+		// Preserve the historical panic value verbatim so any consumer
+		// performing a string-match in a recover() block continues to work.
+		panic("failed to write magic number")
+	}
+	return fw
+}
+
+// NewParquetWriterWithError is identical to [NewParquetWriter] except that
+// it returns an error to the caller if the initial write of the parquet
+// magic header to the underlying sink fails or short-writes. This is the
+// preferred constructor when the sink is backed by a writer that may
+// transiently fail (for example, a Google Cloud Storage resumable upload,
+// an S3 multipart upload, an HDFS client, or any other network-attached
+// writer) where a transient first-write failure should not panic the
+// calling process.
+//
+// If props is nil, then the default Writer Properties will be used. If the
+// key value metadata is not nil, it will be added to the file.
+func NewParquetWriterWithError(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) (*Writer, error) {
 	config := &writerConfig{}
 	for _, o := range opts {
 		o(config)
@@ -87,8 +115,10 @@ func NewParquetWriter(w io.Writer, sc *schema.GroupNode, opts ...WriteOption) *W
 	}
 
 	fw.metadata = *metadata.NewFileMetadataBuilder(fw.Schema, fw.props, config.keyValueMetadata)
-	fw.startFile()
-	return fw
+	if err := fw.startFile(); err != nil {
+		return nil, err
+	}
+	return fw, nil
 }
 
 // NumColumns returns the number of columns to write as defined by the schema.
@@ -167,7 +197,7 @@ func (fw *Writer) appendRowGroup(buffered bool) *rowGroupWriter {
 	return fw.rowGroupWriter
 }
 
-func (fw *Writer) startFile() {
+func (fw *Writer) startFile() error {
 	encryptionProps := fw.props.FileEncryptionProperties()
 	magic := magicBytes
 	if encryptionProps != nil {
@@ -199,8 +229,11 @@ func (fw *Writer) startFile() {
 	}
 
 	n, err := fw.sink.Write(magic)
-	if n != 4 || err != nil {
-		panic("failed to write magic number")
+	if err != nil {
+		return fmt.Errorf("parquet: failed to write magic number: %w", err)
+	}
+	if n != len(magic) {
+		return fmt.Errorf("parquet: short write of magic number: wrote %d of %d bytes", n, len(magic))
 	}
 
 	if fw.props.PageIndexEnabled() {
@@ -209,6 +242,7 @@ func (fw *Writer) startFile() {
 			Encryptor: fw.fileEncryptor,
 		}
 	}
+	return nil
 }
 
 func (fw *Writer) writePageIndex() {

--- a/parquet/file/file_writer_test.go
+++ b/parquet/file/file_writer_test.go
@@ -18,7 +18,9 @@ package file_test
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
+	"io"
 	"math"
 	"reflect"
 	"slices"
@@ -1323,4 +1325,100 @@ func TestBufferedStreamDictionaryCompressed(t *testing.T) {
 	for i := 0; i < numValues; i++ {
 		assert.Equal(t, int32(i), readValues[i])
 	}
+}
+
+// flakyMagicSink models a transient I/O failure on the first Write call,
+// the failure mode reported in apache/arrow-go#820 for cloud-storage upload
+// writers. Set firstErr to fail with that error on the first Write, or set
+// short to true to return n=len(p)-1, nil on the first Write. All subsequent
+// writes succeed and forward to the embedded buffer.
+type flakyMagicSink struct {
+	buf      bytes.Buffer
+	writes   int
+	firstErr error
+	short    bool
+}
+
+func (f *flakyMagicSink) Write(p []byte) (int, error) {
+	f.writes++
+	if f.writes == 1 {
+		if f.firstErr != nil {
+			return 0, f.firstErr
+		}
+		if f.short {
+			return len(p) - 1, nil
+		}
+	}
+	return f.buf.Write(p)
+}
+
+func newSingleColumnSchema(t *testing.T) *schema.GroupNode {
+	t.Helper()
+	fields := schema.FieldList{schema.NewInt32Node("col", parquet.Repetitions.Required, 1)}
+	sc, err := schema.NewGroupNode("schema", parquet.Repetitions.Required, fields, 0)
+	require.NoError(t, err)
+	return sc
+}
+
+func TestNewParquetWriterWithError_Success(t *testing.T) {
+	var buf bytes.Buffer
+	writer, err := file.NewParquetWriterWithError(&buf, newSingleColumnSchema(t))
+	require.NoError(t, err)
+	require.NotNil(t, writer)
+	require.NoError(t, writer.Close())
+}
+
+func TestNewParquetWriterWithError_FirstWriteFails(t *testing.T) {
+	sink := &flakyMagicSink{firstErr: io.ErrUnexpectedEOF}
+	writer, err := file.NewParquetWriterWithError(sink, newSingleColumnSchema(t))
+	require.Error(t, err)
+	require.Nil(t, writer)
+	require.True(t, errors.Is(err, io.ErrUnexpectedEOF),
+		"expected returned error to wrap io.ErrUnexpectedEOF, got %v", err)
+}
+
+func TestNewParquetWriterWithError_FirstWriteShortWrites(t *testing.T) {
+	sink := &flakyMagicSink{short: true}
+	writer, err := file.NewParquetWriterWithError(sink, newSingleColumnSchema(t))
+	require.Error(t, err)
+	require.Nil(t, writer)
+	require.Contains(t, err.Error(), "short write of magic number")
+}
+
+func TestNewParquetWriter_PreservesPanicMessage(t *testing.T) {
+	sink := &flakyMagicSink{firstErr: io.ErrUnexpectedEOF}
+
+	defer func() {
+		r := recover()
+		require.NotNil(t, r, "NewParquetWriter should panic on first-write failure")
+		msg, ok := r.(string)
+		require.True(t, ok, "panic value should remain a string for back-compat (got %T)", r)
+		require.Equal(t, "failed to write magic number", msg)
+	}()
+
+	_ = file.NewParquetWriter(sink, newSingleColumnSchema(t))
+	t.Fatalf("expected NewParquetWriter to panic, but it returned normally")
+}
+
+func TestPqarrowNewFileWriter_PropagatesInitError(t *testing.T) {
+	arrowSchema := arrow.NewSchema([]arrow.Field{{Name: "f", Type: arrow.PrimitiveTypes.Int32}}, nil)
+	sink := &flakyMagicSink{firstErr: io.ErrUnexpectedEOF}
+
+	writer, err := pqarrow.NewFileWriter(arrowSchema, sink,
+		parquet.NewWriterProperties(), pqarrow.DefaultWriterProps())
+	require.Error(t, err)
+	require.Nil(t, writer)
+	require.True(t, errors.Is(err, io.ErrUnexpectedEOF),
+		"expected returned error to wrap io.ErrUnexpectedEOF, got %v", err)
+}
+
+func TestPqarrowNewFileWriter_PropagatesShortWrite(t *testing.T) {
+	arrowSchema := arrow.NewSchema([]arrow.Field{{Name: "f", Type: arrow.PrimitiveTypes.Int32}}, nil)
+	sink := &flakyMagicSink{short: true}
+
+	writer, err := pqarrow.NewFileWriter(arrowSchema, sink,
+		parquet.NewWriterProperties(), pqarrow.DefaultWriterProps())
+	require.Error(t, err)
+	require.Nil(t, writer)
+	require.Contains(t, err.Error(), "short write of magic number")
 }

--- a/parquet/pqarrow/file_writer.go
+++ b/parquet/pqarrow/file_writer.go
@@ -163,7 +163,10 @@ func NewFileWriter(arrschema *arrow.Schema, w io.Writer, props *parquet.WriterPr
 	}
 
 	schemaNode := pqschema.Root()
-	baseWriter := file.NewParquetWriter(w, schemaNode, file.WithWriterProps(props), file.WithWriteMetadata(meta))
+	baseWriter, err := file.NewParquetWriterWithError(w, schemaNode, file.WithWriterProps(props), file.WithWriteMetadata(meta))
+	if err != nil {
+		return nil, err
+	}
 
 	manifest, err := NewSchemaManifest(pqschema, nil, &ArrowReadProperties{})
 	if err != nil {

